bindings/rust/s2n-tls/src/raw/connection.rs

```diff
@@ -362,10 +362,37 @@ impl Connection {
     pub fn negotiate(&mut self) -> Poll<Result<&mut Self, Error>> {
         let mut blocked = s2n_blocked_status::NOT_BLOCKED;
 
-        match unsafe { s2n_negotiate(self.connection.as_ptr(), &mut blocked).into_result() } {
-            Ok(_) => Ok(self).into(),
-            Err(err) if err.is_retryable() => Poll::Pending,
-            Err(err) => Err(err).into(),
+        unsafe {
+            s2n_negotiate(self.connection.as_ptr(), &mut blocked)
+                .into_poll()
+                .map_ok(|_| self)
+        }
+    }
+
+    pub fn send(&mut self, buf: &[u8]) -> Poll<Result<usize, Error>> {
+        let mut blocked = s2n_blocked_status::NOT_BLOCKED;
+        let buf_len: isize = buf.len().try_into().map_err(|_| Error::InvalidInput)?;
+        let buf_ptr = buf.as_ptr() as *const ::libc::c_void;
+        unsafe { s2n_send(self.connection.as_ptr(), buf_ptr, buf_len, &mut blocked).into_poll() }
+    }
+
+    pub fn recv(&mut self, buf: &mut [u8]) -> Poll<Result<usize, Error>> {
+        let mut blocked = s2n_blocked_status::NOT_BLOCKED;
+        let buf_len: isize = buf.len().try_into().map_err(|_| Error::InvalidInput)?;
+        let buf_ptr = buf.as_ptr() as *mut ::libc::c_void;
+        unsafe { s2n_recv(self.connection.as_ptr(), buf_ptr, buf_len, &mut blocked).into_poll() }
+    }
+
+    pub fn flush(&mut self) -> Poll<Result<&mut Self, Error>> {
+        self.send(&[0; 0]).map_ok(|_| self)
```

Comment:
I don't think this actually forces the underlying stream to flush its data, right? I think it would be better to call flush directly on the stream at the higher level.

Reply:
It does not. It just flushes any record s2n-tls has buffered. We should probably flush s2n-tls AND the underlying stream.